Fixes #3079 - Adding display fields to diff view headers in review process

alexcottner opened the pull request:
Fixes #3079 - Adding displayFields properties and values to record headers in diff view.
Added unit test to check

![image](https://github.com/Kinto/kinto-admin/assets/148472676/adedb13d-ee4b-461c-9533-672c44d484ba)

**Note to reviewers:** I am very open to formatting suggestions for making this more readable.

Screenshot with long properties in the header bar:
![image](https://github.com/Kinto/kinto-admin/assets/148472676/bc5dc83b-0afb-4808-8b9b-ac77418f438e)

leplatrem commented on src/components/signoff/SimpleReview/index.tsx:
```diff
@@ -158,6 +158,7 @@ export default function SimpleReview({
         </div>
       );
     }
+    console.log(collection);
```
leftover ;)

alexcottner replied:
Fixed

leplatrem reviewed (commented):
How does it look with long field values? (overflow)

Another approach could be to force the inclusion of lines that contain the display fields. For example, like in pseudo-code
```
rows = diff(r1, r2);
for field in displayFields:
    if field not in rows:
         rows.prepend(field, (r1 || r2)[field])
```

leplatrem commented on src/components/signoff/SimpleReview/PerRecordDiffView.tsx:
```diff
@@ -160,6 +166,34 @@ function DiffLabel({ changeType }: { changeType: ChangeType }) {
+  for (let f of displayFields) {
+    fields.push(
+      <span>
+        <label>{f}:</label> {(target || source)[f] || "undefined"}
```
Since we support dotted field names to target nested objects (eg. `"book.author.name"`), we should use [`renderDisplayField()`](https://github.com/Kinto/kinto-admin/blob/f201c095c81331f295ac988227bbb88446f3f364/src/utils.tsx#L196C17-L196C35) here.

alexcottner replied:
Oh, hadn't considered nested properties being displayed and I didn't write code to compensate for that. Updating to use `renderDisplayField` which also takes care of the `undefined` issue in a consistent way.

leplatrem commented:
Since we are diffing serialized JSON, my suggestion to force certain fields to appear in the diff is a challenging to implement (I just read the source of `utils.diffJson()`), so let's go with this approach.

Before approving I'd like to make sure it renders well when the list of displayField is very long, or contains very long values. And also make sure we support the dotted field notation (see `handleNestedDisplayField()`)

alexcottner commented:
Switched to use `renderDisplayField` as suggested, this keeps us consistent.

As for long headers, it was just wrapping (screenshot below). I'm gonna say this is functional but not pretty, so I adjusted styling to give it a max-width and truncate after that (2nd screenshot below).

How it would have looked:
![image](https://github.com/Kinto/kinto-admin/assets/148472676/6a54f851-41cb-43c5-aa04-5fd1e53a22f1)

How it looks now with truncating:
![image](https://github.com/Kinto/kinto-admin/assets/148472676/70680bf9-91fc-451b-9c44-e241845a9f7d)

leplatrem commented on src/components/signoff/SimpleReview/PerRecordDiffView.tsx:
```diff
@@ -160,6 +166,34 @@ function DiffLabel({ changeType }: { changeType: ChangeType }) {
+  for (let f of displayFields) {
+    fields.push(
+      <span>
+        <label>{f}:</label> {renderDisplayField(target || source, f)}
```
I have a doubt about this. `target` would be `undefined` when record is deleted?  Would it be a tombstone? (eg. `{id: "xxx", deleted: true}`)

alexcottner replied:
If the record was deleted, target would be undefined (or at the very least nullish) so we would render what the source object had for the property.

alexcottner replied:
Example screenshot
![image](https://github.com/Kinto/kinto-admin/assets/148472676/e258a4af-1609-4432-8d35-3bae780e61ac)

leplatrem commented on test/components/signoff/SimpleReview/PerRecordDiffView_test.jsx:
```diff
@@ -135,3 +136,61 @@ describe("findChangeTypes", () => {
+        // prop3 intentionally undefined
+      },
+      displayFields: [ "prop1", "prop2", "prop3"]
+    })).toBe("prop1: val3prop2: val4prop3: <unknown>id: foo");
```
Wouldn't we want `prop3: prevVal` if undefined in target? (genuine question, not sure)
```js
const merged = Object.keys(target).reduce((acc, k) => {
       if (target[k]) { acc[k] = target[k] }
       return acc;
   }, source
);
...
renderDisplayField(merged, f);
```

alexcottner replied:
I thought about that and I don't think so, we can easily coalesce to the property rather than the object to do that (or do a little spread magic). This way we're keeping our behavior consistent with the list view by always rendering the current property values, even when undefined. 

We can always switch this later if the users prefer something else.

leplatrem reviewed (approved):
Looks good. Just one doubt about labels when record is deleted or when property is removed from new record

grahamalama commented:
(Dismissed my review request since it was approved by @leplatrem ✅)
